### ISSUE TITLE
Failure to get inventory of other node via nodemeta for certain coordinates

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#5894]
It seems that there are problem when you try to access an inventory wie `nodemeta` in the `list` formspec element if it is across certain boundaries.

It takes a little bit longer to explain.

Consider this example mod:

```
minetest.register_node("test_chest:chest", {
	description = "Test Chest",
	tiles = {"default_chest_top.png", "default_chest_top.png", "default_cobble.png",
		"default_chest_side.png", "default_chest_side.png", "default_chest_front.png"},
	stack_max = 64,
	groups = {dig_immediate=2},
	is_ground_content = false,
	on_construct = function(pos)
		local meta = minetest.get_meta(pos)
		local inv = meta:get_inventory()
		inv:set_size("main", 8*3)
		inv:set_list("main", {})
	end,

	on_rightclick = function(pos, node, clicker)
		local pos_other = vector.add(pos, {x=1, y=0, z=0})

		minetest.show_formspec(clicker:get_player_name(),
		"test_chest_"..pos.x.."_"..pos.y.."_"..pos.z,

		"size[8,11.5]"..
		"list[nodemeta:"..pos.x..","..pos.y..","..pos.z..";main;0,0.5;8,3;]"..
		"list[nodemeta:"..pos_other.x..","..pos_other.y..","..pos_other.z..";main;0,3.7;8,3;]"..
		"list[current_player;main;0,7.5;8,4]"..
		"listring[current_player;main]"..
		"listring[nodemeta:"..pos.x..","..pos.y..","..pos.z..";main]"..
		"listring[current_player;main]"..
		"listring[nodemeta:"..pos_other.x..","..pos_other.y..","..pos_other.z..";main]")
	end,
})
```

This mod should work in all subgames, but Minetset Game is preferred.

This adds a chest-like node (`test_chest:chest`) with its own inventory. It is some kind of dumbed-down double chest (a chest which occupies 2 nodes on the horizontal plane with double the inventory size).
In the formspec string, two node inventories are attempted to be accessed:
1) The inventory of the node you rightclicked
2) The inventory of the node right from that one (1 to the +X direction, where the cobble texture is), if it exists

So, if you place two of these along the X axis and you rightclick the left one (the one with the lower X coordinate), you should see both chest inventories (2 groups of 3*8). This is intended. If there is no other chest right of the chest, an error is displayed. Also intended.

**However**, if you place a chest pair at any of the following X coordinates, and then open the chest with the lower X coordinate *first*, the other chest inventory is not found, and an error message is displayed. X coordinates:

* (…)
* (31, 32)
* (15, 16)
* (-1, 0)
* (-17, -16)
* (-33, -32)
* (-49, -48)
* (-65, -64)
* (…)
* and so on …

Interestingly, if you rightclick the other chest again, then the first chest, the inventoriy is correctly displayed.

This is very weird, I strongly suspect a bug in Minetest here, as the mod code is extremely simple.

How to reproduce the error:
* Place chest at X=-1
* Place another chest at X=0 (Y and Z must be the same as the first chest)
* Rightclick the first chest
* You see only the inventory of the chest itself, but not the one to the right
* You can click on the empty part of the formspec to make error messages appear. They claim that no inventory was found at the position right of the chest

How to reproduce the happy case:
* Place chest at X=20
* Place another chest at X=19 (Y and Z must be the same as the first chest)
* Rightclick the first chest
* You see both inventories

---

Also try experimenting with other values for `pos_other` by increaing the `x` value, so that the chet tries to find another chest 2 spaces apart, etc. It seems whenever the other chest is beyond any of the above “borders”, the chest can not be found.

You can do the same for Z and Y, it will be the same, just for a different axis.

Interesting observation: If you rightclicked the right chest BEFORE rightclicking the left chest, the error does not occour, even at the “magic numbers”.

This bug has been found by researching a bug in MineClone 2: https://github.com/Wuzzy2/MineClone2-Bugs/issues/162
This is probably the reason why double chests in MineClone 2 are sometimes broken.

Tested in: 9875193d8f7082979702512a247d9472a3345c8c and 0.4.15.

**Comment by Wuzzy2** [comment]
I just made a very interesting observation: If you call `meta:set_string` right after you got the `meta` object in `on_construct`, the bug does not occour.

Replace `on_construct` above with this code:

	on_construct = function(pos)
		local meta = minetest.get_meta(pos)
                meta:set_string("some_string", "some_other_string")
		local inv = meta:get_inventory()
		inv:set_size("main", 8*3)
		inv:set_list("main", {})
	end,

Hmm …
Apparently Minetest has problems if the metadata has not been touched after a node has been placed. Strange.

**Comment by DS-Minetest** [comment]
@Wuzzy2 Afaik `on_construct` the node is not yet placed. => You set the meta of air. And when the air is replaced with the node, the meta is also deleted.

**Comment by Ekdohibs** [comment]
The coordinates you give show that this bug is a mapblock boundary problem. I however don't understand why crossing a mapblock boundary like that creates a bug sometimes, and sometimes not. My best guess would be that the client has not been sent the metadata inside the corresponding mapblock, but I didn't investigate very much more than that.

**Comment by Ekdohibs** [comment]
I finally found why the node metadata is not correctly set: in some cases (maybe due to packet reordering? I am not sure), the client first receives the data for the whole block, including the metadata, and only then the AddNode command. That second command overwrites the node, which was already correct, but also deletes its metadata, leaving the block with incorrect metadata until it is sent again.

**Comment by v-rob** [comment]
This is not a formspec issue, but appears to be a general issue with metadata and possibly network code.  Is this bug still present?

**Comment by Wuzzy2** [comment]
I've just tried to reproduce this bug in 5.9.1, as described in the steps to reproduce above, and in both the error case and the happy case, it works now. The first chest displays the inventory of both nodes.

This bug report is very old, the first time I did a test was apparently in 0.4.15 (according to the description).

I think it's safe to close this bug report. Probably it was fixed in the meanwhile.
